compiler/rustc_trait_selection/src/solve/opaques.rs

```diff
@@ -0,0 +1,62 @@
+use rustc_middle::traits::query::NoSolution;
+use rustc_middle::traits::solve::{Certainty, Goal, QueryResult};
+use rustc_middle::traits::Reveal;
+use rustc_middle::ty;
+use rustc_middle::ty::util::NotUniqueParam;
+
+use super::{EvalCtxt, SolverMode};
+
+impl<'tcx> EvalCtxt<'_, 'tcx> {
+    pub(super) fn normalize_opaque_type(
+        &mut self,
+        goal: Goal<'tcx, ty::ProjectionPredicate<'tcx>>,
+    ) -> QueryResult<'tcx> {
+        let tcx = self.tcx();
+        let opaque_ty = goal.predicate.projection_ty;
+        let expected = goal.predicate.term.ty().expect("no such thing as an opaque const");
+
+        match goal.param_env.reveal() {
+            Reveal::UserFacing => match self.solver_mode() {
+                SolverMode::Normal => {
+                    if !self.can_define_opaque_ty(opaque_ty.def_id) {
+                        return Err(NoSolution);
+                    }
+                    // FIXME: This may have issues when the substs contain aliases...
+                    match self.tcx().uses_unique_placeholders(opaque_ty.substs) {
+                        Err(NotUniqueParam::NotParam(param)) if param.is_non_region_infer() => {
+                            return self.evaluate_added_goals_and_make_canonical_response(
+                                Certainty::AMBIGUOUS,
+                            );
+                        }
+                        Err(_) => {
+                            return Err(NoSolution);
+                        }
+                        Ok(()) => {}
+                    }
+                    // Prefer opaques registered already.
+                    let matches =
+                        self.unify_existing_opaque_tys(goal.param_env, opaque_ty, expected);
+                    if !matches.is_empty() {
+                        if let Some(response) = self.try_merge_responses(&matches) {
+                            return Ok(response);
+                        } else {
+                            return self.flounder(&matches);
+                        }
+                    }
+                    // Otherwise, define a new opaque type
+                    let opaque_ty = tcx.mk_opaque(opaque_ty.def_id, opaque_ty.substs);
+                    self.register_opaque_ty(expected, opaque_ty, goal.param_env)?;
+                    self.evaluate_added_goals_and_make_canonical_response(Certainty::Yes)
+                }
+                SolverMode::Coherence => {
+                    self.evaluate_added_goals_and_make_canonical_response(Certainty::AMBIGUOUS)
+                }
+            },
+            Reveal::All => {
+                let actual = tcx.type_of(opaque_ty.def_id).subst(tcx, opaque_ty.substs);
```

Comment:
try add a debug assertion that predefined opaques are empty

Reply:
:s 

I would need to make the inner infcx reachable from this module, or at least expose a new `EvalCtxt` helper here. I don't think it's worth just yet, so I'll leave a fixme.